File: swara-learn/src/main/java/com/swara/ml/markov/DiscreteMarkovChain.java
package com.swara.ml.markov;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.stream.IntStream;

import com.google.common.base.Preconditions;
import com.swara.ml.common.Trie;

/**
 * A discrete markov chain is a unsupervised random process that undergoes transitions from one
 * state to another. Markov chains are trained on state sequences, from which they approximate the
 * state transition distribution. This transition distribution has the Markov property; the
 * probability of a state transition depends only on the 'order' previous states.
 */
public class DiscreteMarkovChain<T> implements MarkovModel<T> {

    private final Trie<T, Integer> markov;
    private final int order;

    public DiscreteMarkovChain(int order, Comparator<T> comparator) {
        Preconditions.checkArgument(order > 0);
        this.markov = new Trie<>(comparator);
        this.order = order;
    }

    @Override
    public void train(List<T> sequence) {
        // Slides an 'order' length window across the specified input sequence and records a mapping
        // between each 'order' length state sequence and the state that immediately follows it.
        IntStream.range(0, sequence.size() - this.order).parallel().forEach(i -> {
            final List<T> sub = sequence.subList(i, i + this.order + 1);
            this.markov.put(sub, v -> v == null ? 1 : ++v);
        });
    }

    @Override
    public Iterator<T> generate() {
        return new StateIterator(null);
    }

    /**
     * An infinite, lazy iterator over the Markov state space. Every state iterator operates
     * independently, and, therefore, will produce different results. Only iterators with the same
     * seed and the same source of randomness will produce identical results.
     */
    private final class StateIterator implements Iterator<T> {

        private final Random rng;
        private List<T> state;

        public StateIterator(List<T> seed) {
            this.state = seed;
            this.rng = new Random();
        }

        @Override
        public boolean hasNext() {
            return true;
        }

        @Override
        @SuppressWarnings("unchecked")
        public T next() {
            // Calculate a random, valid current state if one does no exist.
            while (this.state == null || markov.get(this.state) == null) {
                Trie<T, Integer> node = markov;
                this.state = new ArrayList<>();
                for (int i = 0; i < order; i++) {
                    final List<Trie<T, Integer>> next = node.children();
                    node = next.get(this.rng.nextInt(next.size()));
                    this.state.add(node.key());
                }
            }

            // Randomly select a valid next state for the given seed.
            final Trie<T, Integer> current = markov.get(this.state);
            final Iterator<Trie<T, Integer>> iter = current.children().iterator();

            Trie<T, Integer> next = null;
            long random = this.rng.nextInt(current.value());
            while (random >= 0 && iter.hasNext()) {
                next = iter.next();
                random -= next.value();
            }

            // Update the seed and return the selected new state.
            this.state.remove(0);
            this.state.add(next.key());
            return next.key();
        }

    }

}

File: swara-learn/src/main/scala/com/swara/learn/markov/DiscreteMarkovChain.scala
package com.swara.learn.markov

import com.swara.learn.{Model, Unsupervised}
import com.swara.learn.common.Trie
import net.jcip.annotations.ThreadSafe
import scala.util.Random

/**
 * A discrete markov chain is a unsupervised random process that undergoes transitions from one
 * state to another. Discrete markov chains are a special case of a family of stochastic models
 * known as markov models, which are used to model random processes in which future states depend
 * only on the current state and not on any prior events. More formally, for discrete sequences,
 * the Markov property requires that the P(x_{n} | x_{n-1}, ... , x_{0}) = P(x_{n} | x_{n-1}).
 * Markov chains are trained on state sequences, from which they approximate the aforementioned
 * state transition probability distribution.
 *
 * @tparam T Type of states.
 */
@ThreadSafe
class DiscreteMarkovChain[T](order: Int) extends Model[Seq[T], T] with Unsupervised[Seq[T]] {

  private[this] val transitions = Trie.empty[T, Int]

  /**
   * Markov models are trained on state sequences, from which they approximate the underlying
   * state transition probability distribution. Implementation slides an 'order + 1' length window
   * across the specified state sequence in parallel and records a mapping between each 'order'
   * length state sequence and the state that immediately follows it.
   *
   * @param example Sequence of states to train on.
   */
  override def train(example: Seq[T]): Unit =
    example.par.iterator
      .sliding(this.order + 1, 1)
      .foreach(seq => this.transitions.put(seq.toList, (suffix, prev) => (suffix, prev) match {
        case (_, None) => Some(1)
        case (_, Some(value)) => Some(value + 1)
      }))

  /**
   * Returns the next state. The next state is chosen using the underlying transition probabilities
   * that define the model. Markov models are inherently non-deterministic; however, they rely on
   * structured randomness to select the most likely next state for any given state sequence. Throws
   * a Runtime Exception if the markov chain has not yet been trained, because all models must be
   * trained before they may be used for predictive tasks.
   *
   * @param states Sequence of states.
   * @return Random next state.
   */
  override def predict(states: Seq[T]): T = {
    val cur = this.transitions.closest(states.toList)
    var num = Random.nextInt(cur.value.getOrElse(0))
    cur.children.dropWhile { child =>
      num -= child.value.getOrElse(0)
      num >= 0
    }.next.symbol.get
  }
  
  /**
   * Markov chains can use the learned state transition probability distribution to generate an
   * arbitrary state sequence. Returns an infinite iterator; note, an iterator is preferable to a
   * stream because streams store all previously computed values in memory.
   *
   * @return An infinite iterator over states.
   */
  def generate(seed: Seq[T] = Seq.empty): Iterator[T] = {
    val state = seed.take(this.order).toBuffer
    Iterator.continually {
      val next = predict(state.toList)
      state.takeRight(this.order - 1) :+ next
      next
    }
  }

}
